crates/aranya-metrics/src/export/prometheus.rs

```
@@ -0,0 +1,186 @@
+use std::{net::SocketAddr, num::NonZeroU32, path::PathBuf, time::Duration};
+
+use anyhow::{Context as _, Result};
+use metrics_exporter_prometheus::PrometheusBuilder;
+use metrics_util::MetricKindMask;
+use serde::Deserialize;
+use tracing::{info, warn};
+use url::Url;
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(default)]
+pub struct PrometheusConfig {
+    /// The current mode that the Prometheus exporter is operating as.
+    mode: PrometheusMode,
+
+    /// Sets the idle timeout for metrics.
+    idle_timeout: (MetricsKindMask, Option<Duration>),
+    /// Sets whether a unit suffix is added to metric names.
+    enable_unit_suffix: bool,
+
+    /// Allows only certain addresses to access the scrape endpoint.
+    allowed_addresses: Option<Vec<String>>,
+    /// Sets the number of quantiles when rendering histograms.
+    quantiles: Option<Vec<f64>>,
+    /// Sets the buckets to use when rendering histograms.
+    buckets: Option<Vec<f64>>,
+    /// Sets the "width" of each bucket when using summaries.
+    bucket_duration: Option<Duration>,
+    /// Sets the number of buckets kept in memory at one time.
+    bucket_count: Option<NonZeroU32>,
+    /// Sets the interval that the upkeep task runs at.
+    upkeep_timeout: Option<Duration>,
+    /// Sets global labels that are applied to all metrics.
+    global_labels: Option<Vec<(String, String)>>,
+}
+
+impl PrometheusConfig {
+    fn format_push_gateway_url(base_url: &str, job_name: &str) -> String {
+        match Url::parse(base_url) {
+            Ok(mut url) => {
+                url.set_path(&format!("/metrics/job/{job_name}"));
+                url.set_query(None);
+                url.set_fragment(None);
+                url.to_string()
+            }
+            Err(_) => {
+                warn!("Failed to parse push gateway URL `{base_url}`");
```

This method should probably return an error if the `base_url` parameter can't be parsed.